File: HW7/AbstractBoard.java
import java.util.Scanner;
import java.io.*;

public abstract class AbstractBoard {
	protected int row;
	protected int column;
	protected String last_move;
	protected int empty_i;
	protected int empty_j;
	protected int wall_number;
	static int numberOfMove = 0;
	static int numberOfBoard = 0;
	
	public AbstractBoard() {
		row = 0;
		column = 0;
		wall_number = 0;
		last_move = "s";
	}

	public AbstractBoard(int row, int column) {
		this.row = row;
		this.column = column;
		wall_number = 0;
		last_move = "s";
	}
	
	/**
	 * That method returns the Board's row number.
	 * @return Board's row number.
	 */
	public int get_row() {
		return row;
	}
	
	/**
	 * That method returns the Board's column number.
	 * @return Board's column number.
	 */
	public int get_column() {
		return column;	
	}
	
	/**
	 * That method returns the Board's empty block's x axis index.
	 * @return Board's empty block's x axis index
	 */
	public int get_empty_i() {
		return empty_i;	
	}
	
	/**
	 * That method returns the Board's empty block's y axis index.
	 * @return Board's empty block's y axis index
	 */
	public int get_empty_j() {
		return empty_j;
	}
	
	
	/**
	 * That method returns the Number of Move.
	 * @return Returns Number of Move
	 */
	public static int numberOfMoves() {
		return numberOfMove;
	}
	
	/**
	 * That method returns the Number of Board.
	 * @return Returns Number of Board
	 */
	public static int numberOfBoards() {
		return numberOfBoard;
	}
	
	/**
	 * That method returns the last move of the game.
	 * @return Returns the last move of the game.
	 */
	public String lastMove() {
		return last_move;
	}
	
	/**
	 * That method assign the parameter row to Board's row number and the parameter column to Board's column number.
	 * @param row  Board's row number
	 * @param column Board's column number
	 */
	public void setSize(int row,int column) {
		this.row = row;
		this.column = column;
	}
	
	/**
	 * That method assign the parameter empty_j to Board's empty block's x axis index.
	 * @param empty_i  Board's empty block's x axis index.
	 */
	public void set_empty_i(int empty_i) {
		this.empty_i = empty_i;
	}
		
	/**
	 * That method assign the parameter empty_j to Board's empty block's y axis index.
	 * @param empty_j Board's empty block's y axis index.
	 */
	public void set_empty_j(int empty_j) {
		this.empty_j = empty_j;
	}
	
	/**
	 * That method checks if there is a same number while setting random board.
	 * @param array the Board which we will check
	 * @param counter counter block number which is row * counter
	 * @param random_number  Number which we will be check
	 * @return if there is same number fail and return 0. if there isn't same number it will be success and returns 1.
	 */
	public int control_random(int array[],int counter,int random_number){ 
		int i;
		for(i = 0 ; i < counter ; i++){
			if(array[i] == random_number)
				return 0; // fail
		}
		return 1;//success
	}

	/**
	 * That method find the row when reading a board from a file.
	 * @param fileName txt file name where we read the board.
	 */
	public void findRow(String fileName) {
		try {
			int counter = 0;
			FileInputStream file = new FileInputStream(fileName);       
			Scanner sc = new Scanner(file); 
			while(sc.hasNextLine())  
			{  
				sc.nextLine();
				counter++;//returns the line that was skipped  
			}  
			sc.close();     //closes the scanner  	
			row = counter;
		}
		catch(FileNotFoundException e) {
			e.printStackTrace();
		}
	}

	/**
	 * That method find the column when reading a board from a file.
	 * @param fileName txt File name where we read the board
	 */
	public void findColumn(String fileName) {
		try {
			FileInputStream file = new FileInputStream(fileName);       
			Scanner input = new Scanner(file); 
			    int counter = 0;
			    wall_number = 0;
			    while (input.hasNext()) {
			      String str = input.next();
			      if(str == "00")
			      	wall_number++;
			      counter++;
			    }
			    input.close(); 
			    column = counter / row;
		}
		catch(FileNotFoundException e) {
			e.printStackTrace();
		}
	}

	/**
	 * That method print the board.
	 */
	public abstract void print();
	
	/**
	 * the blank tiles moves left.
	 */
	public abstract void move_left();
	
	/**
	 * the blank tiles moves right.
	 */
	public abstract void move_right();
	
	/**
	 * the blank tiles moves up.
	 */
	public abstract void move_up();
	
	/**
	 * the blank tiles moves down.
	 */
	public abstract void move_down();
	
	/**
	 * This is the character that the empty box goes this direction. If the parameter is L then the blank tiles moves left. If the parameter is R then the blank tiles moves right. If the parameter is U then the blank tiles moves up, If the parameter is D then the blank tiles moves down
	 * @param direction Direction which we move to
	 */
	public abstract void move(char direction);
	
	/**
	 * This method check the board is finished or continue.
	 * @return if board is finished, method return true,otherwise false.
	 */
	public abstract boolean isSolved();
	
	/**
	 * This method init randomly the board.
	 */
	public abstract void initRandomly();
	
	/**
	 * This method reads the board from txt file which is given as a parameter.
	 * @param fileName txt file which we read from.
	 */
	public abstract void readFromFile(String fileName);
	
	/**
	 * This method writes the board to the txt file which is given as a parameter.
	 * @param fileName txt file name which we will write.
	 */
	public abstract void writeToFile(String fileName);
	
	/**
	 * This method returns board's value on given indexes.
	 * @param i defines x axis index.
	 * @param j  defines y axis index.
	 * @return board's value on given indexes.
	 */
	public abstract int cell(int i,int j);
	
	/**
	 *produces the board as string
	 */
	public abstract String toString();
	
	/**
	 *This method controlled the equivalence of Boards. Board1 is our object, Board2 is given as a parameter. if they are equal returns true,otherwise false.
	 */
	public abstract boolean equals(Object obj);

}

File: HW5/171044003/abstract_board.cpp
#include <iostream>
#include <string>
#include "abstract_board.h"

using std::cout;
using std::cin;
using std::endl;
using std::string;


namespace puzzleOfElif{

	int AbstractBoard::numberOfMove = 0;
	int AbstractBoard::numberOfBoard = 0;

	AbstractBoard::AbstractBoard():row(0),column(0){
		wall_number = 0;
		last_move = "s";
	}
	
	AbstractBoard::AbstractBoard(int b_row,int b_column):row(b_row),column(b_column){
		wall_number = 0;
		last_move = "s";
	}

	AbstractBoard::AbstractBoard(const AbstractBoard& object){ // copy consturctor
		row = object.row;
		column = object.column;
		empty_i = object.empty_i;
		empty_j = object.empty_j;
		last_move = object.last_move;
		wall_number = object.wall_number;
		numberOfMove = object.numberOfMove;
	}

	int AbstractBoard::get_row() const {return row;}
	
	int AbstractBoard::get_column()const {return column;}
	
	int AbstractBoard::get_empty_i()const {return empty_i;}
	
	int AbstractBoard::get_empty_j()const {return empty_j;}
	
	int AbstractBoard::numberOfMoves(){return numberOfMove;}
	
	int AbstractBoard::numberOfBoards(){return numberOfBoard;}
	
	string AbstractBoard::lastMove()const {return last_move;}
	
	void AbstractBoard::setSize(int b_row,int b_column){
		row = b_row;
		column = b_column;
	}

	void AbstractBoard::set_empty_i(int b_empty_i){
		empty_i = b_empty_i;
	}
	
	void AbstractBoard::set_empty_j(int b_empty_j){
		empty_j = b_empty_j;
	}

	void AbstractBoard::menu(){
		int i,j;
		char choice;
		string str;
		last_move = "s"; // 0:left 1:right 2:up 3:down

		do{
				cout << "L:	moves the empty cell left if there is room" << endl;
				cout << "R:	moves the empty cell right if there is room" << endl;
				cout << "U:	moves the empty cell up if there is room" << endl;
				cout << "D:	moves the empty cell down if there is room" << endl;	
				cout << "E:	Asks a file name and saves the current board configuration as a loadable shape file." << endl;
				cout << "O:	Asks a file name and loads the current board configuration from the shape file." << endl;
			
			cout << "choice:" ;
			cin >> choice;
			switch(choice){
				case 'l':
				case 'L':
						move(choice);
						cout << endl;
						print();
						cout << "numberOfMove: " << numberOfMove << endl;
						if(isSolved() == true){
							cout << "Game is solved" << endl;
							exit(1);
						}
						break;
				
				case 'r':
				case 'R':
						move(choice);
						cout << endl;
						print();	
						cout << "numberOfMove: " << numberOfMove << endl;
						if(isSolved()==true){
							cout << "Game is solved" << endl;
							exit(1);
						}
						break;
				
				case 'u':
				case 'U':
						move(choice);
						cout << endl;
						print();
						cout << "numberOfMove: " << numberOfMove << endl;
						if(isSolved()==true){
							cout << "Game is solved" << endl;
							exit(1);
						}
						break;
					
				case 'D':
				case 'd':
						move(choice);
						cout << endl;
						print();
						cout << "numberOfMove: " << numberOfMove << endl;
						if(isSolved()==true){
							cout << "Game is solved" << endl;
							exit(1);
						}
						break;
			
				case 'E':
				case 'e': 
						cout << "Name of the new file: ";
						cin >> str;
						writeToFile(str); 
					break;
				
				case 'O':
				case 'o':
						cout << "Name of the new file: " ;
						cin >> str;
						readFromFile(str);
						print();
						
					break;
				
				case 'q':
				case 'Q':
					break;

				default :
					break;

			}
		}while(choice != 'q' && choice != 'Q');
	}

	int AbstractBoard::stoi_converter(string text){
		char temp;	
		if(text[1] == '\0'){
			temp = text[0];
			text[1] = temp;
			text[0] = 48;
		}
		
		if((text[0] >= 48 && text[0] <= 57) && (text[1] >= 48 && text[1] <= 57))
			return (text[0] - 48) * 10 +(text[1] - 48);
	}

	int	AbstractBoard::control_random(int array[],int counter,int random_number){ // it checks if there is a same number while setting random board.
		int i;
		for(i = 0 ; i < counter ; i++){
			if(array[i] == random_number)
				return 0; // fail
		}
		return 1;//success
	}

}

